### ISSUE TITLE
Issue with fetching/reading events that are out sequence (sequential serial id) but ordered via position

### ISSUE THREAD

**Issue by adriancb** [RailsEventStore/rails_event_store#810]
👋  again team.

Another recent discovery we've made is the reliance on serial ids for conditional reading (`from`, `to`), for <= 1.1.0, and since 1.1.1, batched reading in general.

When we first started using RES, we accidentally emitted certain events out of order. We thought we'd fix our mistake by re-ordering all events within a stream, using position, and that they would then be fetched/read in position order. This is generally the case - unless you're using `from`, `to` or general reads for >= 1.1.1.

I believe the changes at https://github.com/RailsEventStore/rails_event_store/commit/194123b5aedbc39cd5ae371de4b50c652568ef9b#diff-b1f910f1ad306e77c39140dca6493bddbd29d37ec2cccfc173e6bbf86f50fc18R24 has meant that when reading streams containing non-sequentially emitted events, the results contain multiples of the same event (because of multiple batches). The reason, I believe, is that the sequential `id` of the last event isn't the same as the `position`.

It's likely that in <= 1.1.0, using the `from` and `to` condition would've exhibited the same behaviour - we've never used them, so we've been none the wiser.

The documentation for reads (https://railseventstore.org/docs/read/) refers to _event id_ - which we took to mean the _event uuid_. This is obviously not the case and is actually the _sequential event id_ within `event_store_events_in_streams`.

I've included an example that demonstrates the issue with batched reads:

```
event_1 (id: 3, uuid: 'event1', position: 0)
event_2 (id: 4, uuid: 'event1', position: 1)
event_3 (id: 1, uuid: 'event1', position: 2)
event_4 (id: 2, uuid: 'event1', position: 3)
```

In 1.1.1, the result is two batches (due to the serial/sequential offset id calculation):

```
first batch:
event_1 (id: 3, uuid: 'event1', position: 0)
event_2 (id: 4, uuid: 'event1', position: 1)
event_3 (id: 1, uuid: 'event1', position: 2)
event_4 (id: 2, uuid: 'event1', position: 3)

second batch:
event_1 (id: 3, uuid: 'event1', position: 0)
event_2 (id: 4, uuid: 'event1', position: 1)

read output:
event_1
event_2
event_3
event_4
event_1
event_2
```

We'd love some help to understand where to go from here...

**Comment by pawelpacana** [comment]
:wave:

Thank you for a detailed report. 

> We thought we'd fix our mistake by re-ordering all events within a stream, using position, and that they would then be fetched/read in position order. 

Can I ask how did you fix it specifically?

**Comment by pawelpacana** [comment]
Let me answer and clarify bit by bit in the meantime.

1. `from` and `to` 
> The documentation for reads (https://railseventstore.org/docs/read/) refers to event id - which we took to mean the event uuid.   

Correct. Both `from` and `to` expect uuid. Later on these are translated into sequential id within event_store_events_in_streams table in case of a most popular database-backed event repository. 
The reason for this is simply that uuids are not lexicographically sortable. Sequential ids are sortable, allowing to limit dataset by condition.

2. ordering of reads

- reads from given stream are sorted by position in stream first and sequential id second 

This supports all possible values of [expected_version](https://railseventstore.org/docs/v1/expected_version/) where events are strictly ordered or we let the order to be decided by whichever writer succeeds first. The pitfall is when you mix several strategies within a stream, having some events with explicit position and some not.

- reads without specified stream (that is from global stream) are only ordered by sequential id


3. sequential id vs position in stream

Both batch reader (as of v1.1.1) and from-to rely on the monotonicity of `position` and `id` column values. I don't think there is a possibility to append an event to a stream (increasing sequential id) without increasing its expected_version (position).

**Comment by pawelpacana** [comment]
Speaking of the issue you’ve illustrated — we have a [test case](https://github.com/RailsEventStore/rails_event_store/blob/v1.1.1/rails_event_store_active_record/spec/event_repository_spec.rb#L83-L127) in active record event repository where events are re-positioned in a way that seems possible only by manipulating the data outside of RES. We check there that a regular read is still maintaing order by position.

I've adapted this test case to match your example, perform batched read and I can confirm your findings:

```ruby
  specify do
      e1 = Event.create!(
        id: u1 = SecureRandom.uuid,
        data: '{}',
        metadata: '{}',
        event_type: "TestDomainEvent",
      )
      e2 = Event.create!(
        id: u2 = SecureRandom.uuid,
        data: '{}',
        metadata: '{}',
        event_type: "TestDomainEvent",
      )
      e3 = Event.create!(
        id: u3 = SecureRandom.uuid,
        data: '{}',
        metadata: '{}',
        event_type: "TestDomainEvent",
      )
      e4 = Event.create!(
        id: u4 = SecureRandom.uuid,
        data: '{}',
        metadata: '{}',
        event_type: "TestDomainEvent",
      )
      EventInStream.create!(
        stream:   "stream",
        position: 2,
        event_id: e1.id,
      )
      EventInStream.create!(
        stream:   "stream",
        position: 3,
        event_id: e2.id,
      )
      EventInStream.create!(
        stream:   "stream",
        position: 0,
        event_id: e3.id,
      )
      EventInStream.create!(
        stream:   "stream",
        position: 1,
        event_id: e4.id,
      )
      ActiveRecord::Schema.define do
        self.verbose = false
        remove_index :event_store_events_in_streams, [:stream, :position]
      end
      expect(repository.read(specification.stream("stream").in_batches.result).to_a.flatten.map(&:event_id)).to eq([u3, u4, u1, u2])
    end
```

**Comment by pawelpacana** [comment]
> We'd love some help to understand where to go from here...

As long as it is not possible to recreate the non-monotonic `position` only by using RES, I don't see what is there to change in RES itself. New batched reader works within certain assumptions and I yet don't have a reason to believe these assumptions are broken. [Paging by sequential ids](https://developer.wordpress.com/2014/02/14/an-efficient-alternative-to-paging-with-sql-offsets/) as opposed to offset-and-limit gives performance improvement, especially on large datasets.

A solution I see to fix an ordering mistake is to create a new stream, in which events will be linked by its new, correct position. Linking to a new stream results in new entries in `event_store_events_in_stream` in which sequential id and position monotonicity are not in conflict:

```ruby
event_store.read.stream("stream-with-manually-fixed-position-and-sorted-by-position").each.with_index do |event, index|
   event_store.link(event.event_id, stream_name: "new-stream", expected_version: index)
end
```


Perhaps having an option to replace batch reader also solves some problems. I would gladly accept such contribution.

**Comment by adriancb** [comment]
Thanks @pawelpacana - really appreciate the effort (once again) in your reply/ies.

We certainly have a better understanding of the RES design pattern and your replies have validated some of the assumptions we've made to date.

> Can I ask how did you fix it specifically?

We attempted to 'compact' the stream by changing the position (outside of RES). In hindsight, not a great idea...

> The pitfall is when you mix several strategies within a stream, having some events with explicit position and some not.

We faced this issue early on and settled on `auto`, all of the things... There may have been an attempt to backfill the position in some way, which may have led to some of the edge cases we're experiencing.

> Perhaps having an option to replace batch reader also solves some problems. I would gladly accept such contribution.

Given what you've said (and what we've learnt), we'll have a think to see whether our specific edge cases merits a change.

Finally, your suggestion on a path forward is 👍 and we'll make sure to add that (simple) option to our internal documentation (as well as remove any reference to modifying stream/etc outside of RES).
